src/axom/core/ArrayBase.hpp

```diff
@@ -1025,63 +1032,160 @@ struct ArrayOpsBase<T, false>
 };
 
 #if defined(AXOM_USE_GPU) && defined(AXOM_GPUCC) && defined(AXOM_USE_UMPIRE)
+/*!
+ * \name Tag types for device initialization
+ */
+/// @{
+
+/*!
+ * \brief Tag type representing that a type can be initialized on the device.
+ *
+ *  This only applies to types which are trivially device-constructible.
+ */
+struct InitTypeOnDevice
+{ };
+/*!
+ * \brief Tag type representing that a type can be initialized on the device.
+ *
+ *  This applies to types which are not trivially default-constructible, but are
+ *  trivially-copyable; we can construct a default value on the host and copy-
+ *  construct values with it on the device.
+ */
+struct InitTypeOnDeviceWithCopy
+{ };
+/*!
+ * \brief Tag type representing that a type cannot be initialized on the device.
+ */
+struct InitTypeOnHost
+{ };
+
+/*!
+ * \brief Selector type which matches a type to its corresponding initialization
+ *  tag type.
+ */
+template <typename T, typename Enable = void>
+struct DeviceInitTag
+{
+  using Type = InitTypeOnHost;
+};
+
+template <typename T>
+struct DeviceInitTag<T, std::enable_if_t<std::is_trivially_default_constructible<T>::value>>
+{
+  using Type = InitTypeOnDevice;
+};
+
+template <typename T>
+struct DeviceInitTag<T,
+                     std::enable_if_t<!std::is_trivially_default_constructible<T>::value &&
+                                      std::is_default_constructible<T>::value &&
+                                      std::is_trivially_copyable<T>::value>>
+{
+  using Type = InitTypeOnDeviceWithCopy;
+};
+/// @}
+
+template <typename T, OperationSpace SPACE>
+struct DeviceStagingBuffer;
+
 template <typename T>
-struct ArrayOpsBase<T, true>
+struct DeviceStagingBuffer<T, OperationSpace::Device>
+{
+  DeviceStagingBuffer(T* data,
+                      IndexType begin,
+                      IndexType nelems,
+                      bool destroying = false)
+    : m_data(data)
+    , m_begin(begin)
+    , m_num_elems(nelems)
+    , m_is_destroying(destroying)
+  {
+    m_staging_buf = ::operator new(sizeof(T) * nelems);
+    if(destroying)
+    {
+      axom::copy(m_staging_buf, m_data + begin, sizeof(T) * nelems);
+    }
+  }
+
+  DISABLE_COPY_AND_ASSIGNMENT(DeviceStagingBuffer);
+  DISABLE_MOVE_AND_ASSIGNMENT(DeviceStagingBuffer);
+
+  ~DeviceStagingBuffer()
+  {
+    // Copy back staging data to destination buffer.
+    axom::copy(m_data + m_begin, m_staging_buf, m_num_elems * sizeof(T));
+    ::operator delete(m_staging_buf);
+  }
+
+  T* getStagingBuffer() const { return static_cast<T*>(m_staging_buf); }
+
+  void* m_staging_buf;
+  T* m_data;
+  IndexType m_begin;
+  IndexType m_num_elems;
+  bool m_is_destroying;
+};
```

Comment:
Out of curiosity -- does new/delete always work in this context? 
Would people want to use Umpire, e.g. to use memory from an existing pool?

Reply:
I modified this to use Umpire via `axom::allocate()`/`axom::deallocate()`.